src/main/java/carpet/mixins/PistonBlock_movableTEMixin.java

```diff
@@ -0,0 +1,123 @@
+package carpet.mixins;
+
+import carpet.CarpetSettings;
+import carpet.fakes.PistonBlockEntityInterface;
+import com.google.common.collect.Lists;
+import net.minecraft.block.*;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.block.piston.PistonHandler;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.World;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import java.util.List;
+import java.util.Set;
+
+@Mixin(PistonBlock.class)
+public abstract class PistonBlock_movableTEMixin extends FacingBlock
+{
+    protected PistonBlock_movableTEMixin(Settings block$Settings_1)
+    {
+        super(block$Settings_1);
+    }
+    
+    private ThreadLocal<List<BlockEntity>> list1_BlockEntities = new ThreadLocal<>(); //Unneccessary ThreadLocal if client and server use different PistonBlock instances
+    
+    @Inject(method = "isMovable", at = @At(value = "RETURN", ordinal = 3, shift = At.Shift.BEFORE))
+    private static void movableCMD(BlockState blockState_1, World world_1, BlockPos blockPos_1,
+            Direction direction_1, boolean boolean_1, Direction direction_2, CallbackInfoReturnable<Boolean> cir)
+    {
+        Block block_1 = blockState_1.getBlock();
+        //Make CommandBlocks movable, either use instanceof CommandBlock or the 3 cmd block objects,
+        if (CarpetSettings.b_movableTileEntities && block_1 instanceof CommandBlock)
+        {
+            cir.setReturnValue(true);
+        }
+    }
+    
+    private static boolean isPushableTileEntityBlock(Block block)
+    {
+        //Making PISTON_EXTENSION (BlockPistonMoving) pushable would not work as its createNewTileEntity()-method returns null
+        return block != Blocks.ENDER_CHEST && block != Blocks.ENCHANTING_TABLE &&
+                       block != Blocks.END_GATEWAY && block != Blocks.END_PORTAL && block != Blocks.MOVING_PISTON  &&
+                       block != Blocks.SPAWNER;
+    }
+    
+    @Redirect(method = "isMovable", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;hasBlockEntity()Z"))
+    private static boolean ifHasBlockEntity(Block block)
+    {
+        if (!block.hasBlockEntity())
+        {
+            return !true;
```

Comment:
!true? you mean false...

Reply:
This bit was by 2No... i did ask him the same...he said he did so for better readability...and i do agree with him